src/Microsoft.Health.Fhir.TemplateManagement/ArtifactProviders/BlobTemplateProvider.cs

```diff
@@ -0,0 +1,90 @@
+using DotLiquid;
+using Microsoft.Extensions.Caching.Memory;
+using Microsoft.Health.Fhir.Liquid.Converter.Utilities;
+using Microsoft.Health.Fhir.TemplateManagement.Models;
+
+namespace Microsoft.Health.Fhir.TemplateManagement.ArtifactProviders
+{
+    public class BlobTemplateProvider : IConvertDataTemplateProvider
+    {
+        private BlobContainerClient _blobContainerClient;
+
+        private readonly IMemoryCache _templateCache;
+
+        private readonly TemplateCollectionConfiguration _templateCollectionConfiguration;
+
+        private readonly string _blobTemplateCacheKey = "cached-blob-templates";
+
+        private readonly int _segmentSize = 100;
+
+        public BlobTemplateProvider(BlobContainerClient blobContainerClient, IMemoryCache templateCache, TemplateCollectionConfiguration templateConfiguration)
+        {
+            _blobContainerClient = blobContainerClient;
+            _templateCache = templateCache;
+            _templateCollectionConfiguration = templateConfiguration;
+        }
+
+        public async Task<List<Dictionary<string, Template>>> GetTemplateCollectionAsync(CancellationToken cancellationToken = default)
+        {
+            // read templates from cache if available
+            if (_templateCache.TryGetValue(_blobTemplateCacheKey, out var templateCache))
+            {
+                return (List<Dictionary<string, Template>>)templateCache;
+            }
+
+            var templateNames = await ListBlobsFlatListing(_blobContainerClient, _segmentSize, cancellationToken);
+
+            var templates = new Dictionary<string, string>();
+
+            foreach (var templateName in templateNames)
+            {
+                var template = await DownloadBlobToStringAsync(_blobContainerClient, templateName);
+                templates.Add(templateName, template);
+            }
+
+            var parsedtemplates = TemplateUtility.ParseTemplates(templates);
+
+            var templateCollection = new List<Dictionary<string, Template>> { parsedtemplates };
+
+            _templateCache.Set(_blobTemplateCacheKey, templateCollection, _templateCollectionConfiguration.ShortCacheTimeSpan);
+
+            return templateCollection;
+        }
+
+        private static async Task<List<string>> ListBlobsFlatListing(BlobContainerClient blobContainerClient, int? segmentSize, CancellationToken ct)
+        {
+            var blobs = new List<string>();
+
+            var resultSegment = blobContainerClient.GetBlobsAsync()
+                .AsPages(default, segmentSize);
+
+            await foreach (Page<BlobItem> blobPage in resultSegment)
+            {
+                foreach (BlobItem blobItem in blobPage.Values)
+                {
+                    blobs.Add(blobItem.Name);
+                }
+            }
+
+            return blobs;
+        }
+
+        public static async Task<string> DownloadBlobToStringAsync(BlobContainerClient blobContainerClient, string blobName)
```

Comment:
What should our error handing here be? If we encounter some exception when downloading a specific template, I think we should continue to attempt downloading others with a retry policy (with backoff) on this operation.

Reply:
We should be careful of waiting too long with the backoff policy to avoid unnecessary delays in call times but as long as the timeframe is in the milliseconds range I agree.

Reply:
I added some retries for throwing an exception. I think we'll throw for now but I can see us adding more sophisticated logic in the future